osgpkitools/utils.py

```diff
@@ -58,3 +133,133 @@ def safe_write(filename, contents):
     """
     safe_rename(filename)
     atomic_write(filename, contents)
+
+def check_permissions(path):
+    """The function checks for write permissions for the given path to verify if the user has write permissions
+    """
+    if os.access(path, os.W_OK):
+        return
+    else:
+        raise FileWriteException("User does not have appropriate permissions for writing to current directory.")
+
+
+def find_user_cred(usercert=None, userkey=None):
+    """Find a readable user cert/key pair, trying pairs in the following order:
+    1. usercert, userkey
+    2. X509_USER_CERT, X509_USER_KEY environment variables
+    3. '~/.globus/usercert.pem', '~/.globus/userkey.pem'
+    INPUT (optional)
+    usercert: path to user certificate
+    userkey: path to private key of user
+    OUTPUT
+    Paths to the user cert and key
+    """
+    
+    # list of cert/key pairs to try
+    input_pairs = [(usercert, userkey),
+                   ((os.environ.get('X509_USER_CERT'), os.environ.get('X509_USER_KEY'))),
+                   (os.path.expanduser('~/.globus/usercert.pem'), os.path.expanduser('~/.globus/userkey.pem'))]
+    cert_key_pairs = [t for t in input_pairs if None not in t] # remove undefined pairs for an improved err msg below
+
+    # M2Crypto doesn't raise exceptions when encountering missing or unreadable
+    # cert/key pairs so we force the issue
+    for cert, key in cert_key_pairs:
+        try:
+            open(cert, 'r')
+            open(key, 'r')
+            return cert, key
+        except IOError:
+            continue
+    raise IOError("Unable to read the following certificate/key pairs:\n- %s" %
+                  "\n- ".join([", ".join(pair) for pair in cert_key_pairs]))
+
+
+def print_failure_reason_exit(data):
+    """This functions prints the failure reasons and exits"""
+    try:
+        msg = 'The request has failed for the following reason: %s' % \
+        json.loads(data)['detail'].split('--')[1].lstrip()
+    except IndexError:
+        msg = 'The request has failed for the following reason: %s' % json.loads(data)['detail'].lstrip() + \
+              'Status : %s ' % json.loads(data)['status']
+
+    separator = '='*80
+    sys.exit('\n'.join(textwrap.wrap(separator + msg, width=80)))
+
+
+class Cert(object):
+
+    KEY_LENGTH = 2048
+    PUB_EXPONENT = 0x10001
+
+    def __init__(self, common_name, output_dir=None, altnames=None):
+        """Create a certificate request (stored in the x509request attribute) and associated keys (stored in keypair attribute).
+        The caller should use write_pkey to write private key when ready.
+
+        This function accepts the CN and final path for the key as well as optional list of subject alternative names
+        and optional requestor e-mail.  """
+        escaped_common_name = common_name.replace('/', '_') # Remove / from service requests for writing keys
```

Comment:
The comment is referring to the service request that the tool is making so I think it's valid when understood, but I'll change it to just "Remove slashes".

Reply:
If people are requesting common names with slashes in them, they're going to have problems down the line, though.

Reply:
We have seen people use them in service certificates, that's why Jeny put this in.

Reply:
So, InCommon does not allow slashes in the DNs, we decided to use dashes instead. 
Example: /CN=novapro-novagpvm01.fnal.gov 

I think we can get rid of this line.

Reply:
My experience is that InCommon follows the CA/Browser Baseline Requirements, which disallows '/' in the CN field. If we accommodate it, the tool will happily create a CSR/keypair for a certificate that InCommon will not issue, and so the tool will fail on an rejected request. I'm suggesting that instead we get rid of this line and catch improperly formatted CNs when we validate script input.

Reply:
Ok I deleted it.

Now I think the comment used to mean to refer to "service certificates", not "service requests".  InCommon does support service certificates, but Let's Encrypt does not.

Reply:
>  InCommon does support service certificates

This is news to me. Have you guys been able to successfully get service certs from InCommon? My understanding, which is a few months old, is that InCommon CAs follow the CA/Browser Baseline Requirements that has a strict list of allowed characters in the CN field

Reply:
What does a service certificate mean to you?  It is a client certificate not assigned to a particular user.  We can get any certificate that has a CN ending in fnal.gov, so if it doesn't match a host name and it is allowed to be used as a client certificate (which these are), then I think it can be considered a service certificate.  It doesn't have to contain a specific format in the CN field.

Reply:
Ah ok, that's fine then. I was referring to the IGTF definition of service certificates, e.g. `pilot/flock.opensciencegrid.org`, which is what this line was handling.

Reply:
I wasn't aware of that.  Not having the slash shouldn't prevent people from using them though.  Maybe we should push for them to accept slashes to adhere to the IGTF standard; we got them to change some other things in the DN that were against IGTF standards.

Reply:
Sure if you can get them to change it but I think if there's a conflict between the baseline reqs and the IGTF standards, they'll pick the baseline reqs.